File: 1372. Longest ZigZag Path in a Binary Tree.js
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestZigZag = function(root) {
    let res = 0;
    
    function dfs(node, is_left, count) {
        if (node) {
            res = Math.max(res, count);
            if (is_left) {
                dfs(node.left, false, 1);
                dfs(node.right, true, count+1);
            } else {
                dfs(node.right, true, 1);
                dfs(node.left, false, count+1);
            }
        }
    }
    
    dfs(root, true, 0);
    dfs(root, false, 0);
    return res;
};

File: 1372. Longest ZigZag Path in a Binary Tree.java
class Solution {
    public int longestZigZag(TreeNode root) {
        int[] res = new int[1];
        dfs(root, true, 0, res);
        dfs(root, false, 0, res);
        return res[0];
    }
    
    void dfs(TreeNode node, boolean is_left, int count, int[] res) {
        if (node != null) {
            res[0] = Math.max(res[0], count);
            if (is_left) {
                dfs(node.left, false, 1, res);
                dfs(node.right, true, count+1, res);
            } else {
                dfs(node.right, true, 1, res);
                dfs(node.left, false, count+1, res);
            }
        }
    }
}
